File: src/formatter.js
#!/usr/bin/env node
/**
 * Workout Log Formatter
 * Converts JSON records back into concise human-readable strings.
 */

const fs = require('fs');
const readline = require('readline');

// Helper to capitalize words
function capitalize(str) {
  if (!str) return '';
  return str.split(/[_\s]+/).map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
}

// Format a list of sets into a concise string
// e.g. "315x5x3" or "315x5, 315x4" or "20, 20, 25"
function formatSets(sets, unit = 'lb') {
  if (!sets || sets.length === 0) return 'No sets';

  // Group by weight
  const groups = [];
  let currentGroup = null;

  for (const set of sets) {
    const weight = set.weight !== undefined ? set.weight : null;
    const reps = set.reps;
    const failed = set.failed ? ' (fail)' : '';

    if (currentGroup && currentGroup.weight === weight) {
      currentGroup.reps.push(`${reps}${failed}`);
    } else {
      if (currentGroup) groups.push(currentGroup);
      currentGroup = { weight, reps: [`${reps}${failed}`] };
    }
  }
  if (currentGroup) groups.push(currentGroup);

  // Format groups
  return groups.map(g => {
    // Check if all reps are the same
    const allSame = g.reps.every(r => r === g.reps[0]);
    const count = g.reps.length;
    const repStr = allSame && count > 1 ? `${count}x${g.reps[0]}` : g.reps.join(', ');

    if (g.weight !== null) {
      // Strength: "315 lb: 5x5" or just "315x5x5" style?
      // User asked for concise. "315x5x5" is classic.
      // But if reps vary: "315x5,4,3"
      if (allSame && count > 1) {
        return `${g.weight}x${count}x${g.reps[0]}`; // 315x5x5
      } else if (count === 1) {
         return `${g.weight}x${g.reps[0]}`;
      } else {
        return `${g.weight}x[${repStr}]`;
      }
    } else {
      // Bodyweight: "20, 20, 25" or "5x10"
      if (allSame && count > 1) {
        return `${count}x${g.reps[0]} reps`;
      }
      return `${repStr} reps`;
    }
  }).join(', ');
}

function formatRecord(record) {
  const date = record.ts.split('T')[0];
  const name = capitalize(record.exercise || record.modality);
  const notes = record.notes ? ` - "${record.notes}"` : '';
  const rpe = record.rpe ? ` @ RPE ${record.rpe}` : '';

  let summary = '';

  if (record.type === 'cardio') {
    const parts = [];
    if (record.duration_min) parts.push(`${record.duration_min} min`);
    if (record.distance_miles) parts.push(`${record.distance_miles} mi`);
    if (record.speed_mph) parts.push(`${record.speed_mph} mph`);
    if (record.incline_percent) parts.push(`${record.incline_percent}% inc`);
    summary = parts.join(', ');
  } else {
    // Strength / Machine / Bodyweight
    summary = formatSets(record.sets, record.unit);
  }

  return `[${date}] ${name}: ${summary}${rpe}${notes}`;
}

// Export
module.exports = { formatRecord };

// CLI
if (require.main === module) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false
  });

  rl.on('line', (line) => {
    if (!line.trim()) return;
    try {
      const record = JSON.parse(line);
      console.log(formatRecord(record));
    } catch (e) {
      // Ignore invalid JSON lines or print error?
      // Silently ignore to be pipe-friendly
    }
  });
}

File: workout_logger/formatter.py
"""
Workout log formatter - converts JSONL records to human-readable format.
Ported from src/formatter.js
"""

import sys
import json
from typing import Dict, Any, List


def capitalize(s: str) -> str:
    """Capitalize words in a string, handling underscores."""
    if not s:
        return ''
    return ' '.join(word.capitalize() for word in s.replace('_', ' ').split())


def formatSets(sets: List[Dict[str, Any]], unit: str = 'lb') -> str:
    """
    Format a list of sets into a concise string.
    e.g. "315x5x3" or "315x5, 315x4" or "20, 20, 25"
    """
    if not sets or len(sets) == 0:
        return 'No sets'

    # Group by weight
    groups = []
    current_group = None

    for s in sets:
        weight = s.get('weight')
        reps = s.get('reps')
        failed = ' (fail)' if s.get('failed') else ''

        if current_group and current_group['weight'] == weight:
            current_group['reps'].append(f'{reps}{failed}')
        else:
            if current_group:
                groups.append(current_group)
            current_group = {'weight': weight, 'reps': [f'{reps}{failed}']}

    if current_group:
        groups.append(current_group)

    # Format groups
    result = []
    for g in groups:
        # Check if all reps are the same
        all_same = all(r == g['reps'][0] for r in g['reps'])
        count = len(g['reps'])
        rep_str = f"{count}x{g['reps'][0]}" if all_same and count > 1 else ', '.join(g['reps'])

        if g['weight'] is not None:
            # Strength: "315x5x5" or "315x[5,4,3]"
            if all_same and count > 1:
                result.append(f"{int(g['weight'])}x{count}x{g['reps'][0]}")
            elif count == 1:
                result.append(f"{int(g['weight'])}x{g['reps'][0]}")
            else:
                result.append(f"{int(g['weight'])}x[{rep_str}]")
        else:
            # Bodyweight: "20, 20, 25" or "5x10"
            if all_same and count > 1:
                result.append(f"{count}x{g['reps'][0]} reps")
            else:
                result.append(f"{rep_str} reps")

    return ', '.join(result)


def formatRecord(record: Dict[str, Any]) -> str:
    """
    Format a single workout record into human-readable format.

    Examples:
        "[2026-02-16] Squat: 315x5x3 @ RPE 8 - "felt strong""
        "[2026-02-16] Treadmill: 10 min, 3.2 mph, 15% inc"
        "[2026-02-16] Note: "Felt tired today""
    """
    date = record['ts'].split('T')[0]
    name = capitalize(record.get('exercise') or record.get('modality', ''))
    notes = f' - "{record["notes"]}"' if record.get('notes') else ''
    rpe = f' @ RPE {record["rpe"]}' if record.get('rpe') else ''

    if record['type'] == 'note':
        return f'[{date}] Note: "{record.get("notes", "")}"'

    if record['type'] == 'cardio':
        parts = []
        if record.get('duration_min'):
            parts.append(f"{record['duration_min']} min")
        if record.get('distance_miles'):
            parts.append(f"{record['distance_miles']} mi")
        if record.get('speed_mph'):
            parts.append(f"{record['speed_mph']} mph")
        if record.get('incline_percent'):
            parts.append(f"{record['incline_percent']}% inc")
        summary = ', '.join(parts)
    else:
        # Strength / Machine / Bodyweight
        summary = formatSets(record.get('sets', []), record.get('unit', 'lb'))

    return f'[{date}] {name}: {summary}{rpe}{notes}'


def main():
    """CLI entry point - read JSONL from stdin and output formatted records."""
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        try:
            record = json.loads(line)
            print(formatRecord(record))
        except json.JSONDecodeError:
            # Silently ignore invalid JSON lines to be pipe-friendly
            continue
        except Exception as e:
            # Optionally log errors to stderr
            print(f"Error formatting record: {e}", file=sys.stderr)
            continue


if __name__ == '__main__':
    main()
